kornia/geometry/epipolar/projection.py

```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
 
     return K @ Rt
 
+def KRt_from_projection(P: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    r"""Get the intrinsics, rotation-matrix and the camera-center.
+
+    This function decomposes the Projection matrix into Camera-Matrix, Rotation Matrix and Translation vector.
+
+
+    Args:
+        P (torch.Tensor): the projection matrix with shape :math:`(B, 3, 4)`.
+
+    Returns:
+        torch.Tensor: The Camera matrix with shape :math:`(B, 3, 3)`.
+        torch.Tensor: The Rotation matrix with shape :math:`(B, 3, 3)`.
+        torch.Tensor: the Translation vector with shape :math:`(B, 3)`.
+
+    """
+
+    assert P.shape[-2:] == (3, 4), "P must be of shape [B, 3, 4]"
+
+    submat_3x3 = P[:, 0:3, 0:3]
+    last_column = P[:, 0:3, 3].unsqueeze(-1)
+    t = torch.matmul(-torch.inverse(submat_3x3), last_column).squeeze(-1)
+
+    # Trick to turn QR-decomposition into RQ-decomposition
+    reverse = torch.Tensor([[0, 0, 1], [0, 1, 0], [1, 0, 0]]).unsqueeze(0)
+    submat_3x3 = torch.matmul(reverse, submat_3x3).permute(0, 2, 1)
+    ortho_mat, upper_mat = torch.qr(submat_3x3)
+    ortho_mat = torch.matmul(reverse, ortho_mat.permute(0, 2, 1))
+    upper_mat = torch.matmul(reverse, torch.matmul(upper_mat.permute(0, 2, 1), reverse))
+
+    # Turning the `upper_mat's` diagonal elements to positive.
+    diagonals = torch.diagonal(upper_mat, dim1=-2, dim2=-1)
+    signs = diagonals / diagonals.abs()
+    signs_mat = torch.diag_embed(signs)
+
+    K = torch.matmul(upper_mat, signs_mat)
+    R = torch.matmul(signs_mat, ortho_mat)
+    return K, R, t
```

Comment:
what formulation are you following for this method ? I would refer in the code.
Besides, I'd suggest to follow opencv sfm since this module is inspired on that. See: https://github.com/opencv/opencv_contrib/blob/ab2e83248fa1783d7c2055ef280d403ffcf968c8/modules/sfm/src/libmv_light/libmv/multiview/projection.cc#L32
Open to others alternatives though.

Reply:
I have referred to the following links for my implementation.
1. [Turning a QR decomposition to RQ decomposition](https://math.stackexchange.com/questions/1640695/rq-decomposition/1640762).

2. [Forcing all elements of Camera matrix to be positive](http://ksimek.github.io/2012/08/14/decompose/).

Currently, my implementation forces all the elements of `K` to be positive.
But, in-case of OpenCV's [decomposeProjectionMatrix](https://github.com/opencv/opencv/blob/60354e39018ac82d230d0d021cdc77343fc84ee8/modules/calib3d/src/calibration.cpp#L3271) implementation, sometimes the `last column of K` and `last row of R` turns out to be  `-ve`  wrt my code.

Have a look at the picture below.
![diff-opencv-torch](https://user-images.githubusercontent.com/35777220/91661474-f23c0f00-eaf9-11ea-802b-415ef56ad9c4.png)

This is due to the fact that inverting a specific column and corresponding row leads to the same solution.

I can change my current implementation to replicate the results as in OpenCV.

Reply:
I'd prefer to replicate opencv results for convenience.

Reply:
Sure, I'll update the code.

Reply:
I will be implementing the algorithm as written in https://github.com/opencv/opencv_contrib/blob/ab2e83248fa1783d7c2055ef280d403ffcf968c8/modules/sfm/src/libmv_light/libmv/multiview/projection.cc#L32

Some of the test-cases did not pass in the previous algorithm, hence I had to write code as per the link pasted above.